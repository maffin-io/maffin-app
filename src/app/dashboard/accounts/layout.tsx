import React from 'react';

export const metadata = {
  title: 'Home',
};

export default function AccountsLayout({
  children,
}: React.PropsWithChildren): JSX.Element {
  return (
    <div>
      {children}
    </div>
  );
}
