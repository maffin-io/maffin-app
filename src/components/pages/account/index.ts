export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
