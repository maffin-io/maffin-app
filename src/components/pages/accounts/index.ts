export { default as AccountsTable } from './AccountsTable';
export { default as MonthlyTotalHistogram } from './MonthlyTotalHistogram';
export { default as NetWorthPie } from './NetWorthPie';
export { default as LatestTransactions } from './LatestTransactions';
