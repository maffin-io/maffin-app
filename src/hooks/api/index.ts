import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
