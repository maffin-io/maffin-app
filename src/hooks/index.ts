export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
