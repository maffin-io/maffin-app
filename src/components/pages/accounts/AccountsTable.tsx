import React from 'react';
import { DateTime } from 'luxon';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';
import { Tooltip } from 'react-tooltip';

import Money from '@/book/Money';
import Table from '@/components/Table';
import type { AccountsMap } from '@/types/book';
import { MonthlyTotals } from '@/lib/queries';
import { Account } from '@/book/entities';
import {
  isInvestment,
  isAsset,
  isLiability,
} from '@/book/helpers/accountType';
import { useAccountsTotals, useAccounts } from '@/hooks/api';
import mapAccounts from '@/helpers/mapAccounts';

export type AccountsTableProps = {
  selectedDate?: DateTime,
  isExpanded?: boolean,
};

type AccountsTableRow = {
  account: Account,
  total: Money,
  leaves: AccountsTableRow[],
};

export default function AccountsTable(
  {
    selectedDate = DateTime.now(),
    isExpanded = false,
  }: AccountsTableProps,
): JSX.Element {
  const { data } = useAccounts();
  const { data: monthlyTotals } = useAccountsTotals();

  const accounts = mapAccounts(data);
  const tree = getTreeTotals(accounts.type_root, accounts, monthlyTotals || {}, selectedDate);

  return (
    <Table<AccountsTableRow>
      id="accounts-table"
      columns={columns}
      data={tree.leaves}
      initialState={{
        sorting: [{ id: 'total', desc: true }],
      }}
      showHeader={false}
      tdClassName="p-2"
      getSubRows={row => row.leaves}
      isExpanded={isExpanded}
    />
  );
}

function getTreeTotals(
  current: Account,
  accounts: AccountsMap,
  monthlyTotals: MonthlyTotals,
  selectedDate: DateTime,
): AccountsTableRow {
  const leaves: AccountsTableRow[] = [];
  current.childrenIds.forEach(childId => {
    const childAccount = accounts[childId];
    if (!childAccount.hidden && childAccount.parentId === current.guid) {
      leaves.push(getTreeTotals(childAccount, accounts, monthlyTotals, selectedDate));
    }
  });

  let accountTotal = (monthlyTotals[current.guid] || {})[selectedDate.toFormat('MM/yyyy')]
    || new Money(0, current.commodity?.mnemonic || '');
  if (!isAsset(current) && !isLiability(current)) {
    accountTotal = Object.entries(monthlyTotals[current.guid] || {}).reduce(
      (total, [monthYear, amount]) => {
        if (DateTime.fromFormat(monthYear, 'MM/yyyy') <= selectedDate) {
          return total.add(amount);
        }
        return total;
      },
      new Money(0, current.commodity?.mnemonic || ''),
    );
  }

  return {
    account: current,
    total: current.type === 'INCOME' || isLiability(current) ? accountTotal.multiply(-1) : accountTotal,
    leaves,
  };
}

const columns: ColumnDef<AccountsTableRow>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-50" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-50" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-50" />
          </button>
        )}
        {
          (
            row.original.account.placeholder
            && (
              <span
                data-tooltip-id={row.original.account.guid}
                className={classNames('badge cursor-default', {
                  success: row.original.account.type === 'INCOME',
                  danger: row.original.account.type === 'EXPENSE',
                  info: isAsset(row.original.account),
                  warning: isLiability(row.original.account),
                  misc: isInvestment(row.original.account),
                })}
              >
                {row.original.account.name}
              </span>
            )
          ) || (
            <Link
              data-tooltip-id={row.original.account.guid}
              className={classNames('badge hover:text-slate-300', {
                success: row.original.account.type === 'INCOME',
                danger: row.original.account.type === 'EXPENSE',
                info: isAsset(row.original.account),
                warning: isLiability(row.original.account),
                misc: isInvestment(row.original.account),
              })}
              href={`/dashboard/accounts/${row.original.account.guid}`}
            >
              {row.original.account.name}
            </Link>
          )
        }

        {
          row.original.account.description
          && (
            <Tooltip
              id={row.original.account.guid}
              className="tooltip"
              disableStyleInjection
            >
              {row.original.account.description}
            </Tooltip>
          )
        }
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTableRow) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
