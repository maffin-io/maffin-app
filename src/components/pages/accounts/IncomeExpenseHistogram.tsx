import React from 'react';
import { DateTime, Interval } from 'luxon';
import type { ChartDataset } from 'chart.js';

import Bar from '@/components/charts/Bar';
import { moneyToString } from '@/helpers/number';
import { useAccountsMonthlyTotal, useMainCurrency } from '@/hooks/api';

export type IncomeExpenseHistogramProps = {
  selectedDate?: DateTime,
};

export default function IncomeExpenseHistogram({
  selectedDate = DateTime.now(),
}: IncomeExpenseHistogramProps): JSX.Element {
  const interval = Interval.fromDateTimes(
    selectedDate.minus({ months: 6 }).startOf('month'),
    selectedDate,
  );
  const { data: monthlyTotals } = useAccountsMonthlyTotal(interval);
  const incomeSeries = monthlyTotals?.income;
  const expenseSeries = monthlyTotals?.expense;

  const { data: currency } = useMainCurrency();
  const unit = currency?.mnemonic || '';

  const dates = interval.splitBy({ month: 1 }).map(d => (d.start as DateTime).startOf('month'));
  const datasets: ChartDataset<'bar'>[] = [
    {
      label: 'Income',
      data: [],
      backgroundColor: '#22C55E',
    },
    {
      label: 'Expenses',
      data: [],
      backgroundColor: '#EF4444',
    },
    {
      label: 'Savings',
      data: [],
      backgroundColor: '#06B6D4',
      datalabels: {
        anchor: 'end',
        display: true,
        formatter: (value) => moneyToString(value, unit),
        align: 'end',
        backgroundColor: '#06B6D4FF',
        borderRadius: 5,
        color: '#FFF',
      },
    },
  ];

  dates.forEach(date => {
    const monthYear = (date as DateTime).toFormat('MM/yyyy');
    const incomeAmount = (incomeSeries?.[monthYear]?.toNumber() || 0);
    const expenseAmount = (expenseSeries?.[monthYear]?.toNumber() || 0);
    const netProfit = -incomeAmount - expenseAmount;

    datasets[0].data.push(-incomeAmount || 0);
    datasets[1].data.push(-expenseAmount || 0);
    datasets[2].data.push(netProfit || 0);
  });

  return (
    <>
      <Bar
        height="400"
        data={{
          labels: dates,
          datasets,
        }}
        options={{
          layout: {
            padding: {
              right: 15,
            },
          },
          maintainAspectRatio: false,
          interaction: {
            mode: 'index',
          },
          scales: {
            x: {
              type: 'time',
              time: {
                unit: 'month',
                tooltipFormat: 'MMMM yyyy',
                displayFormats: {
                  month: 'MMM-yy',
                },
              },
              grid: {
                display: false,
              },
              ticks: {
                align: 'center',
              },
            },
            y: {
              grace: 1,
              position: 'left',
              border: {
                display: false,
              },
              ticks: {
                maxTicksLimit: 10,
                callback: (value) => moneyToString(value as number, unit),
              },
            },
          },
          plugins: {
            title: {
              display: true,
              text: 'Monthly Savings',
              align: 'start',
              padding: {
                top: 0,
                bottom: 30,
              },
              font: {
                size: 18,
              },
            },
            legend: {
              onClick: () => {},
              position: 'bottom',
              labels: {
                usePointStyle: true,
                pointStyle: 'circle',
                boxHeight: 8,
                boxWidth: 8,
              },
            },
            tooltip: {
              backgroundColor: '#323b44',
              callbacks: {
                label: (ctx) => `${ctx.dataset.label}: ${moneyToString(Number(ctx.raw), unit)}`,
                labelColor: (ctx) => ({
                  borderColor: '#323b44',
                  backgroundColor: ctx.dataset.backgroundColor as string,
                  borderWidth: 3,
                  borderRadius: 2,
                }),
              },
            },
          },
        }}
      />
      <span />
    </>
  );
}
