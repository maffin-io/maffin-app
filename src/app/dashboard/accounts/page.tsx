'use client';

import React from 'react';
import { DateTime } from 'luxon';
import type { SWRResponse } from 'swr';

import type { Account } from '@/book/entities';
import Money from '@/book/Money';
import AccountsTable from '@/components/AccountsTable';
import AddAccountButton from '@/components/buttons/AddAccountButton';
import NetWorthPie from '@/components/pages/accounts/NetWorthPie';
import { PriceDBMap } from '@/book/prices';
import DateRangeInput from '@/components/DateRangeInput';
import { useApi } from '@/hooks';
import type { AccountsTree } from '@/types/accounts';

export default function AccountsPage(): JSX.Element {
  const { data: earliestDate } = useApi('/api/start-date');
  let { data: accounts } = useApi('/api/accounts/splits') as SWRResponse<Account[]>;
  let { data: todayPrices } = useApi('/api/prices/today');

  const [selectedDate, setSelectedDate] = React.useState(DateTime.now());

  accounts = accounts || [];
  todayPrices = todayPrices || new PriceDBMap();

  let tree: AccountsTree = {
    account: {},
    total: new Money(0, 'EUR'),
    children: [],
  };
  const root = accounts.find(a => a.type === 'ROOT');
  if (root && !todayPrices.isEmpty) {
    tree = buildNestedRows(root, accounts, todayPrices, selectedDate);
  }

  return (
    <>
      <div className="flex items-center pb-4">
        <span className="text-xl font-medium">
          Your finances
        </span>
        <span className="ml-auto mr-3">
          <DateRangeInput
            earliestDate={earliestDate}
            dateRange={{
              start: selectedDate,
              end: selectedDate,
            }}
            onChange={(value: { start: DateTime, end: DateTime }) => setSelectedDate(value.start)}
            asSingle
          />
        </span>
        <div>
          <AddAccountButton />
        </div>
      </div>
      <div className="grid grid-cols-12 items-start items-top pb-4">
        <div className="col-span-3 p-4 mr-4 rounded-sm bg-gunmetal-700">
          <NetWorthPie tree={tree} />
        </div>
        <div className="col-span-3 p-4 mr-4 rounded-sm bg-gunmetal-700">
          <AccountsTable tree={tree} />
        </div>
      </div>
    </>
  );
}

/**
 * This function recursively goes through all the accounts so it can build the tree
 * of accounts containing total and subrows so they can be rendered in react table.
 *
 * Note that it interacts with PriceDB so prices can be aggregated to the parent's
 * total in order to display total amounts of assets, income, etc.
 */
function buildNestedRows(
  current: Account,
  accounts: Account[],
  todayQuotes: PriceDBMap,
  date: DateTime,
): AccountsTree {
  const { childrenIds } = current;
  let total = current.getTotal(date);
  let children: AccountsTree[] = [];

  if (current.type === 'ROOT') {
    children = childrenIds.map(childId => {
      const child = accounts.find(a => a.guid === childId) as Account;
      const childTree = buildNestedRows(child, accounts, todayQuotes, date);
      return childTree;
    });
  } else {
    const commodity = current.commodity.mnemonic;
    if (['STOCK', 'MUTUAL'].includes(current.type)) {
      const price = todayQuotes.getStockPrice(
        current.commodity.mnemonic,
        DateTime.now(),
      );
      total = total.convert(price.currency.mnemonic, price.value);
    }
    children = childrenIds.map(childId => {
      const child = accounts.find(a => a.guid === childId) as Account;
      const childTree = buildNestedRows(child, accounts, todayQuotes, date);
      let childTotal = childTree.total;
      if (childTotal.currency !== commodity) {
        const price = todayQuotes.getPrice(
          childTotal.currency,
          commodity,
          DateTime.now(),
        );
        childTotal = childTotal.convert(commodity, price.value);
      }
      total = total.add(childTotal);
      return childTree;
    });
  }

  return {
    account: current,
    total,
    children,
  };
}
