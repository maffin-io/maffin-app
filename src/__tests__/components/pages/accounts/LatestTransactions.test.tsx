import React from 'react';
import { render, screen } from '@testing-library/react';
import { DateTime } from 'luxon';
import type { UseQueryResult } from '@tanstack/react-query';

import { LatestTransactions } from '@/components/pages/accounts';
import * as apiHook from '@/hooks/api';
import type { Transaction } from '@/book/entities';

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

describe('LatestTransactions', () => {
  beforeEach(() => {
    jest.spyOn(apiHook, 'useLatestTxs').mockReturnValue({ data: undefined } as UseQueryResult<Transaction[]>);
  });

  it('renders with empty txs', () => {
    const { container } = render(<LatestTransactions />);
    expect(container).toMatchSnapshot();
  });

  it('renders as expected with txs', () => {
    jest.spyOn(apiHook, 'useLatestTxs').mockReturnValue({
      data: [
        {
          description: 'tx 1',
          date: DateTime.fromISO('2023-01-01'),
          currency: {
            mnemonic: 'EUR',
          },
          splits: [
            {
              account: {
                guid: 'account1',
                name: 'Account 1',
                type: 'ASSET',
                commodity: {
                  mnemonic: 'EUR',
                },
              },
              quantity: 100,
            },
          ],
        },
        {
          description: 'tx 2',
          date: DateTime.fromISO('2023-01-02'),
          currency: {
            mnemonic: 'EUR',
          },
          splits: [
            {
              account: {
                guid: 'account2',
                name: 'Account 2',
                type: 'ASSET',
                commodity: {
                  mnemonic: 'SGD',
                },
              },
              quantity: -100,
            },
          ],
        },
      ],
    } as UseQueryResult<Transaction[]>);
    const { container } = render(<LatestTransactions />);

    screen.getByText('Account 1');
    screen.getByText('Account 2');
    expect(container).toMatchSnapshot();
  });

  it('picks the correct split when investment', () => {
    jest.spyOn(apiHook, 'useLatestTxs').mockReturnValue({
      data: [
        {
          description: 'tx 1',
          date: DateTime.fromISO('2023-01-01'),
          currency: {
            mnemonic: 'EUR',
          },
          splits: [
            {
              account: {
                guid: 'account2',
                name: 'Account 2',
                type: 'INVESTMENT',
                commodity: {
                  mnemonic: 'EUR',
                },
              },
              quantity: 100,
            },
            {
              account: {
                guid: 'account1',
                name: 'Account 1',
                type: 'ASSET',
                commodity: {
                  mnemonic: 'EUR',
                },
              },
              quantity: 100,
            },
          ],
        },
      ],
    } as UseQueryResult<Transaction[]>);
    const { container } = render(<LatestTransactions />);

    screen.getByText('Account 1');
    expect(container).toMatchSnapshot();
  });
});
