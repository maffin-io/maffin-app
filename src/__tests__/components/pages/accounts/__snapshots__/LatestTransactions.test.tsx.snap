// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LatestTransactions picks fallback split on split event transaction 1`] = `
<div>
  <div>
    <p>
      Latest transactions
    </p>
    <div
      class="text-sm py-3"
    >
      <span
        class="flex items-center"
      >
        <div
          class="mr-2"
        >
          <svg
            class="text-xl text-success"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 8.414V18h2V8.414l4.293 4.293 1.414-1.414L12 4.586l-6.707 6.707 1.414 1.414z"
            />
          </svg>
        </div>
        <div
          class="w-full"
        >
          <span
            class="inline-flex items-center"
          >
            <svg
              class="mr-1"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 11h2v2H7zm0 4h2v2H7zm4-4h2v2h-2zm0 4h2v2h-2zm4-4h2v2h-2zm0 4h2v2h-2z"
              />
              <path
                d="M5 22h14c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2h-2V2h-2v2H9V2H7v2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2zM19 8l.001 12H5V8h14z"
              />
            </svg>
            01 Jan
             - 
            tx 1
          </span>
          <div
            class="flex items-center"
          >
            <span
              class="text-success"
            >
              €100.00
            </span>
            <a
              class="ml-auto text-xs badge hover:text-slate-300 bg-misc bg-info"
              href="/dashboard/accounts/account2"
            >
              Account 2
            </a>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`LatestTransactions picks the correct split when investment 1`] = `
<div>
  <div>
    <p>
      Latest transactions
    </p>
    <div
      class="text-sm py-3"
    >
      <span
        class="flex items-center"
      >
        <div
          class="mr-2"
        >
          <svg
            class="text-xl text-success"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 8.414V18h2V8.414l4.293 4.293 1.414-1.414L12 4.586l-6.707 6.707 1.414 1.414z"
            />
          </svg>
        </div>
        <div
          class="w-full"
        >
          <span
            class="inline-flex items-center"
          >
            <svg
              class="mr-1"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 11h2v2H7zm0 4h2v2H7zm4-4h2v2h-2zm0 4h2v2h-2zm4-4h2v2h-2zm0 4h2v2h-2z"
              />
              <path
                d="M5 22h14c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2h-2V2h-2v2H9V2H7v2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2zM19 8l.001 12H5V8h14z"
              />
            </svg>
            01 Jan
             - 
            tx 1
          </span>
          <div
            class="flex items-center"
          >
            <span
              class="text-success"
            >
              €100.00
            </span>
            <a
              class="ml-auto text-xs badge hover:text-slate-300 bg-info"
              href="/dashboard/accounts/account1"
            >
              Account 1
            </a>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`LatestTransactions renders as expected with txs 1`] = `
<div>
  <div>
    <p>
      Latest transactions
    </p>
    <div
      class="text-sm py-3"
    >
      <span
        class="flex items-center"
      >
        <div
          class="mr-2"
        >
          <svg
            class="text-xl text-success"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 8.414V18h2V8.414l4.293 4.293 1.414-1.414L12 4.586l-6.707 6.707 1.414 1.414z"
            />
          </svg>
        </div>
        <div
          class="w-full"
        >
          <span
            class="inline-flex items-center"
          >
            <svg
              class="mr-1"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 11h2v2H7zm0 4h2v2H7zm4-4h2v2h-2zm0 4h2v2h-2zm4-4h2v2h-2zm0 4h2v2h-2z"
              />
              <path
                d="M5 22h14c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2h-2V2h-2v2H9V2H7v2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2zM19 8l.001 12H5V8h14z"
              />
            </svg>
            01 Jan
             - 
            tx 1
          </span>
          <div
            class="flex items-center"
          >
            <span
              class="text-success"
            >
              €100.00
            </span>
            <a
              class="ml-auto text-xs badge hover:text-slate-300 bg-info"
              href="/dashboard/accounts/account1"
            >
              Account 1
            </a>
          </div>
        </div>
      </span>
    </div>
    <div
      class="text-sm py-3"
    >
      <span
        class="flex items-center"
      >
        <div
          class="mr-2"
        >
          <svg
            class="text-xl text-danger"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18.707 12.707-1.414-1.414L13 15.586V6h-2v9.586l-4.293-4.293-1.414 1.414L12 19.414z"
            />
          </svg>
        </div>
        <div
          class="w-full"
        >
          <span
            class="inline-flex items-center"
          >
            <svg
              class="mr-1"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 11h2v2H7zm0 4h2v2H7zm4-4h2v2h-2zm0 4h2v2h-2zm4-4h2v2h-2zm0 4h2v2h-2z"
              />
              <path
                d="M5 22h14c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2h-2V2h-2v2H9V2H7v2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2zM19 8l.001 12H5V8h14z"
              />
            </svg>
            02 Jan
             - 
            tx 2
          </span>
          <div
            class="flex items-center"
          >
            <span
              class="text-danger"
            >
              -SGD 100.00
            </span>
            <a
              class="ml-auto text-xs badge hover:text-slate-300 bg-info"
              href="/dashboard/accounts/account2"
            >
              Account 2
            </a>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`LatestTransactions renders with empty txs 1`] = `
<div>
  <div>
    <p>
      Latest transactions
    </p>
  </div>
</div>
`;
